### PR TITLE
spec: add some methods for HartMask

### PR DESCRIPTION
**Pull request by woshiluo**
- Add `HartIds` structure as iteration of HartMask, also add test for testing it.
- Add `HartMask::insert` and `HartMask::remove`, and add its test.

### CODE REVIEW

File: sbi-spec/src/binary.rs [hunk 1]
```diff
@@ -794,6 +804,20 @@ pub struct HartMask {
     hart_mask_base: usize,
 }
 
+/// Iteration for HartMask, from low to high.
+pub struct HartMaskIter {
```
Comment (luojia65) [inline sbi-spec/src/binary.rs:808]: Please consider derive `Clone`, `Copy`, `Debug`, `Eq`, `PartialEq` and `Hash` for this structure. We exclude `Ord` and `PartialOrd` for that this structure does not define any orderings.

File: sbi-spec/src/binary.rs [hunk 2]
```diff
@@ -829,6 +853,63 @@ impl HartMask {
...
+            },
+            visited_mask: 0,
+        }
+    }
+}
+
+impl Iterator for HartMaskIter {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
```
Comment (luojia65) [inline sbi-spec/src/binary.rs:901]: Should we `#[inline]` this function? We can compare compilation results for functions that include or does not include inline hints to see how they differs.

File: sbi-spec/src/binary.rs (hunk 1 above)
Comment (woshiluo) [reply sbi-spec/src/binary.rs:809]: Derived, thanks for point it out.

File: sbi-spec/src/binary.rs (hunk 2 above)
Comment (woshiluo) [reply sbi-spec/src/binary.rs:904]: With inline hint, it seems even unrolling the loop of `for x in mask.iter()`. 

So I added inline hints here, thanks for this notice.

**Review by luojia65 — APPROVED** [review]
This pull request LGTM

**Pull request merged**